Face blurring

hunminkim98 opened the pull request:
Hello, Dr. David,

I hope you are doing well.

I am sharing the current version of the face blurring script.

https://github.com/user-attachments/assets/9bf11748-8cc1-49c3-91a6-f51636365151

**Key Featrues**
- Face Detection: detects faces using the RTMO with COCO_17 model from rtmlib.
- Face Obscuration: obscures detected faces with either cv2.GaussianBlur or a black shape.
- Configurable Shape: choose between a rectangle (default, faster) or a polygon (closer fit using convex hull).
- Full Face Coverage: estimates forehead and chin keypoints baed on eys and nose, with adjustable parameter (--blur-size: small, medium, large).
- Optimized Performance: Initially, I faced a quite slow speed. So, I implement applying blurring or black shape only within Region Of Interest (approx x5 fater on my end).
- Adjustable Blur: blur intensity can be set "low" or "medium" or "high".

**Outputs**
- Will save a video with obscured faces and added prefix name "blurred_"'
- Optionally saves face keypoints as json files.

Currently, the script processes one video at a time using the -i argument.
I would like to implement batch processing as well and was wondering how best to extend it to cover all levels of Pose2Sim's folder structure.
It would be greatly helpful if you could share any advice on this!

Always, I really appreciate your time and consideration.

Best regards,

HunMin Kim

davidpagnon commented:
Thanks again Hunmin (and Yacine, whose idea it was) for this very nice tool!

I brought a few changes; it looks like a lot but really, they are just details:
- I simplified the call of the RTMO model: we do not need the PoseTracker class, so we can simply invoke the Body one with pose='rtmo'
- I set rtmpose as the default model: since we can run person detection only every few frames, it is much faster (I set det_frequency to 10 by default but I added a parameter in case you want to increade it even more)
- I added an option for the user to choose the backend and device to be used
- Instead of "--no_json", I replaced with the argument "--save_json" as I think people are more likely not to need to save face keypoints. 

Other small changes
- I replaced the dash in the arguments with underscores for better consistency with the other Utilities scripts
- I implicitly declared the arguments in min(): `face_blurring_func(**args)`
- I replaced all occurrences of os.path by Path from pathlib. It does not make any difference, but the code is slightly more condensed

And other modified files:
- I also added face_blurring to the entry points in pyproject.toml, so that it can be called from anywhere (ie, not necessarily from the Utilities folder)
- And I added a line about it in the Readme

Aside from these very minor changes, it is great and is a very useful tool added to the library!
Feel free to test it and amend anything you like, and then we can merge the pull request.

PS: You are right, batch processing all videos in a directory, or accepting several videos as input, could be nice! (but I did not do it)
Maybe a time_range argument, too (not sure about this one)

hunminkim98 commented:
I apologize for the delayed response. I’ve been taking some rest due to a cold.

Thank you very much for sharing such valuable feedback and thoughtful modifications. I’ve learned a lot once again :)

I’ll test the revised version and work on adding the batch processing feature tonight!

hunminkim98 commented:
Hello again,
I've tested your modification and added batch processing.

Your modification worked very well and clearly! Thank you very much for your effort :)

Here are the changes I made:
- Explicit Path conversion: I explicitly converted the Path object to a string using str() before passing it to cv2.VideoCapture.
- Batch processing support: Added functionality to process videos in batch mode. When users provide the -r or --root argument with a root directory, the script will automatically process all videos in the folder except those with filenames ending in _blurred.
- Input validation logic: Added a simple check to ensure that users supply either the -r/--root or -i/--input argument.

davidpagnon commented:
Well I think it is ready to be merged!
I added other video extensions, so that most video types are detected.

Also, right now if a person is not detected on one or several frames, their face won't be blurred, so I added a warning message about it. I was thinking of doing some sort of tracking to make sure they are not lost but it is not always conclusive.

hunminkim98 commented:
Awesome! This is definitely clearer than before.

I appreciate your time and effort!
